### ISSUE TITLE
 include badge number update in the push notification payload details

### ISSUE THREAD

**Issue by petervojtek** [PowerlineApp/powerline-mobile#252]
from #226:

Please include badge number update in the push notification payload details. https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md#badges Badge number is the same as the count of the user's total priority items. If not easy, ignore badge. https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/API.md#pushsetapplicationiconbadgenumbersuccesshandler-errorhandler-count---ios--android-only

**Comment by jterps08** [comment]
Is this fairly easy? What does this involve exactly?

**Comment by igaponov** [comment]
@jterps08 I'm working on it

**Comment by jterps08** [comment]
ok cool - if it was too intense we could just have just delayed it, but it looks like you're done already :-)

**Comment by petervojtek** [comment]
@jterps08 , I am confused  - where do you want the badge count to appear?

the push notification includes the `count` attribute as required by [api doc](https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/API.md#pushsetapplicationiconbadgenumbersuccesshandler-errorhandler-count---ios--android-only)

```
{"message":" commented on your petition","image":"http://powerline-dev.imgix.net/avatars/577cb68e6f276.jpeg?dpr=0.75&h=400&ixlib=php-1.1.0&w=320","title":"Peter10 Is Leader","count":"34","additionalData":{"collapse_key":"do_not_collapse","type":"own-user-petition-commented","foreground":true,"actions":[{"callback":"app.view","icon":"View","title":"View"},{"callback":"app.mute","icon":"Mute","title":"Mute"}],"entity":{"target":{"id":257,"comment_id":200,"label":"petition","type":"user-petition","preview":"test 13"},"id":1879}}} 
```

However I cannot see the badge count to appear either on the push notification, nor on the Powerline app icon in list of apps.

On Android 4.4, I cannot see the badge count anywehre (nor on push notification, nor on icon).
On android 6, I cannot see the badge count on app icon, but the push notification has following weird small circle:

![screenshot_20161003-150409](https://cloud.githubusercontent.com/assets/225506/19065178/478780d6-8a12-11e6-9f7e-88f787e94ef4.png)

**Comment by jterps08** [comment]
Peter

1) Badge that I am referring to is system OS app icon badge number. (e.g. on OS home screen) See: https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/API.md#pushsetapplicationiconbadgenumbersuccesshandler-errorhandler-count---ios--android-only AND https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md#badges

2) The weird little circle is because there are supposed to be two icons displayed. Large icon is avatar, small icon is always the Powerline logo. In example below, large icon is square. Small icon is circular. https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md#images

![image](https://cloud.githubusercontent.com/assets/7582457/19073146/68509b78-8a07-11e6-8603-dba18ccf118c.png)

**Comment by petervojtek** [comment]
ok, so for 1) we are blocked by by #295

**Comment by petervojtek** [comment]
ok, I tried to call `setApplicationIconBadgeNumber` as depicted in the docs, but both on android 4.4 and 6 it returns "Invalid action ". This is unresolved issue [on stackoverflow](http://stackoverflow.com/questions/22055054/phonegap-push-notification-is-there-a-way-to-add-up-badge-number#comment40486507_22055054).

I will keep this open until iOS notifications start to arrive (#295) and then check it on iOS.

**Comment by jterps08** [comment]
I looked at the stackoverflow and that issue seems to be related to 'adding' to the previous badge number amount... I would think that we woul dbe able to pass a count of priority activities for the user from the server to the notification payload. There are no issues in https://github.com/phonegap/phonegap-plugin-push/issues?page=1&q=is%3Aissue+is%3Aopen+badge&utf8=%E2%9C%93 that speak to a problem with this item.

**Comment by petervojtek** [comment]
@jterps08 , the stackoverflow issue deals with the what you describe, but I specifically pointed out this comment:

> "have you got the android setApplicationIconBadgeNumber to actually work?? For me it always returns an error: invalid action".

and that is what I experience as well (it has nothing to do with incrementing the badge count, I call it like `setApplicationIconBadgeNumber(7)`, i.e. I do not try to fetch the previous number or so..

**Comment by jterps08** [comment]
I feel like this is a question for the github repo for the push plugin... It's a very active repo, so I'd be surprised if others did not experience this problem...

**Comment by petervojtek** [comment]
@jterps08 - yes, but that was the moment I realized it will take more than 15 minutes to sort out this issue, and given it is P3, I rather stopped paying effort to it now.

**Comment by petervojtek** [comment]
the badge count via `setApplicationIconBadgeNumber`works well on iOS:

![img_0011](https://cloud.githubusercontent.com/assets/225506/19120134/39f3b654-8b22-11e6-8295-b04d5c5f3223.PNG)

so I will have to resolve it for Android..

**Comment by jterps08** [comment]
Looks good!

**Comment by petervojtek** [comment]
@igaponov , it seems that you do not attach the badge count in iOS push notifications::

```
{"title":"Peter10 Is Leader","sound":"default","image":"http://powerline-dev.imgix.net/avatars/577cb68e6f276.jpeg?dpr=0.75&h=400&ixlib=php-1.1.0&w=320","message":" commented on your post","additionalData":{"type":"own-post-commented","foreground":true,"coldstart":false,"entity":"{\"id\":1947,\"target\":{\"id\":320,\"preview\":\"c13\",\"type\":\"post\",\"label\":\"post\",\"comment_id\":687}}"}}
```

while on android the `count` is there:

```
{"count":"39","image":"http://powerline-dev.imgix.net/avatars/577cb68e6f276.jpeg?dpr=0.75&h=400&ixlib=php-1.1.0&w=320","title":"Peter10 Is Leader","message":" commented on the post you subscribed to","additionalData":{"entity":{"id":1959,"target":{"id":322,"preview":"hi1","type":"post","label":"post","comment_id":694}},"actions":[{"icon":"View","callback":"app.view","title":"View"},{"icon":"Mute","callback":"app.mute","title":"Mute"}],"type":"follow-post-commented","collapse_key":"do_not_collapse","foreground":true}}
notifications.js:29 setApplicationIconBadgeNumber failed: 
```

**Comment by igaponov** [comment]
@petervojtek [here](https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md) I see a paragraph with badges only for Android. iOS has only `"badge": true`

**Comment by petervojtek** [comment]
@jterps08 , on Android the situation is now following:

I managed to resolve the previous problem by updating the push plugin library (the former version did not supported badges on Android).

Now the currently used version of `phonegap-plugin-push`(1.8.2) will successfully proceed `setApplicationIconBadgeNumber`, but the badge number of not displayed near the poweline app icon. I tested that on Android 4.4 and Android 6. 

Even when the badge number will be displayed successfully, I am not sure what is the added value for the user, because there is currently no business logic to downcrease the badge number when user went inside the app and did some actions which downcrease the number of items in priority zone. So the badge number will currently update only with next push notification arrival.

I am not sure how much time you want spend on this feature currently.. apart from debugging the android, this badge counters will need more testing and debugging to make sure they are properly processed disregarding if the powerline app is closed/in background/in foreground (the push notifications behave differently for each of these conditions). 
So we have here three states multiplied by two operating systems.. this issue could easily take several hours to spend on..

**Comment by petervojtek** [comment]
> @petervojtek here I see a paragraph with badges only for Android. iOS has only "badge": true

@igaponov , I do not understand -- can you show me where is it written that badges do not work on iOS? I can successfully display the badge number on iPhone (by using some random number).. so all I need is to put the count anywhere in the notification (at least this works when the mobile app is open when the push notification arrives).

**Comment by jterps08** [comment]
We need badge counters for iOS and Android. They don't need to be perfect. They need to signal the user to come into the app if there's anything in the priority zone. The badge counter should only disappear (=0) if there are no priority items for the user to address. 

Note: If the badge counter cannot zero (as I think you're explaining above Peter) and will have a 1+ number for the time being, that is fine. The point is to get users to open the app. We can fix the badge counter to =0 when the priority zone count = 0 in the future.

**Comment by petervojtek** [comment]
@jterps08 , for Android it seems that it won't be possible (in short amount of time) to have the badge count in every android device.

> Android ("vanilla" android without custom launchers and touch interfaces) does not allow changing of the application icon, because it is sealed in the .apk tightly once the program is compiled. There is no way to change it to a 'drawable' programmatically using standard APIs. You may achieve your goal by using a widget instead of an icon.

[source](http://stackoverflow.com/questions/17565307/how-to-display-count-of-notifications-in-app-launcher-icon)

Now the point is --  the phonegap-plugin-push is using [this tool](https://github.com/leolin310148/ShortcutBadger) to display the badge counts, but it works on:

> An Android library supports badge notification like iOS in Samsung, LG, Sony and HTC launchers.

That is why when i called `phonegap-plugin-push.setApplicationIconBadgeNumber` I have not seen the badge count on the powerline app icon -- because my android 4.4 is emulator's vanilla android (with stock launcher), and my real Android 6 is Lenovo Moto, which is famous for having almost vanilla android with stock android launcher.

in a nutshell, there will be some android users who will see the badge count (actually it will be majority of android powerline users), but we cannot guarantee it for all.

**Comment by jterps08** [comment]
Okay understood - that's fine. Some is better than none :-)
thank you for clarification.

**Comment by petervojtek** [comment]
@igaponov , can you please incude the badge count for iOS also this way (and also keep it please in the `count` attribute inside push message payload as you already do):

```
{
"APNS_SANDBOX":"{\"aps\":{\"alert\":<push msg>, \"badge\":<badge-count-here>}}"
}
```

this way the badge count will update on iOS even when the user won't open the notification.

**Comment by petervojtek** [comment]
its done
